scripts/pretraining/bert/pretraining_utils.py

```diff
@@ -0,0 +1,273 @@
+    dataset_cached : bool, default is False
+        Whether or not to cache last processed dataset.
+        Each processed dataset can only be cached for once.
+        When there is no new available processed dataset to be fetched,
+        we pop a cached processed dataset.
+    num_max_dataset_cached : int, default is 0
+        Maximum number of cached datasets. It is valid only if dataset_cached is True
+    """
+    num_files = len(glob(data))
+    logging.info('%d files are found.', num_files)
+    assert num_files >= num_parts, \
+        'The number of text files must be no less than the number of ' \
+        'workers/partitions (%d). Only %d files at %s are found.'%(num_parts, num_files, data)
+    dataset_params = {'allow_pickle': True}
+    sampler_params = {'batch_size': batch_size, 'shuffle': shuffle, 'num_buckets': num_buckets}
+    dataset_fn = prepare_pretrain_npz_dataset
+    sampler_fn = prepare_pretrain_bucket_sampler
+    pad_val = vocab.pad_id
+    batchify_fn = bf.Tuple(
+        bf.Pad(val=pad_val, round_to=8),  # input_id
+        bf.Pad(val=pad_val),  # masked_id
+        bf.Pad(val=0),  # masked_position
+        bf.Pad(val=0),  # masked_weight
+        bf.Stack(),  # next_sentence_label
+        bf.Pad(val=0, round_to=8),  # segment_id
+        bf.Stack())  # valid_lengths
+    split_sampler = SplitSampler(num_files, num_parts=num_parts,
+                                 part_index=part_idx, repeat=repeat)
+    dataloader = DatasetLoader(data,
+                               file_sampler=split_sampler,
+                               dataset_fn=dataset_fn,
+                               batch_sampler_fn=sampler_fn,
+                               dataset_params=dataset_params,
+                               batch_sampler_params=sampler_params,
+                               batchify_fn=batchify_fn,
+                               num_dataset_workers=num_dataset_workers,
+                               num_batch_workers=num_batch_workers,
+                               pin_memory=False,
+                               circle_length=circle_length,
+                               dataset_cached=dataset_cached,
+                               num_max_dataset_cached=num_max_dataset_cached)
+    return dataloader
+
+
+
+class MaskedAccuracy(EvalMetric):
+    def __init__(self, axis=1, name='masked_accuracy',
+                 output_names=None, label_names=None):
+        super(MaskedAccuracy, self).__init__(name, axis=axis,
+            output_names=output_names, label_names=label_names)
+        self.axis = axis
+
+    def update(self, labels, preds, masks=None):
+        masks = [None] * len(labels) if masks is None else masks
+        for label, pred_label, mask in zip(labels, preds, masks):
+            if pred_label.shape != label.shape:
+                #pred_label = pred_label.argmax(axis=self.axis)
+                pred_label = mx.npx.topk(pred_label.astype('float32', copy=False),
```

Comment:
Have you met some bugs of argmax?

Reply:
`mx.npx.topk` is faster than `mx.np.argmax` about 50 times. The issue (https://github.com/apache/incubator-mxnet/issues/11061) still exists in MXNet 2.0.

```
import time
import mxnet as mx
import numpy as np
import os
tmp = mx.np.random.normal(-1, 1, (64,300000), ctx=mx.gpu())
tic = time.time()
for i in range(20):
   if (i == 5):
        begin = time.time();
   elif (i == 15):
        end = time.time();
   tic = time.time()
   #out = mx.np.argmax(tmp, axis=1)
   out = mx.npx.topk(tmp, k=1, ret_typ='indices', axis=1, dtype=np.int32)
   out.wait_to_read()
   toc = time.time() - tic
   print ("used time %f:"%toc)
avg = (end - begin) / 10
print ("avg time %f"%avg)
```

Reply:
Thanks Ziyue, I can reproduce.... 

@szha Our argmax is too slow...

Reply:
I'm actually not aware of this. Argmax is quite easy to accelerate and we can even call packages like cub.